### PR TITLE
Add support for torch v2.0.0

### PR DESCRIPTION
**Pull request by jkulhanek**
This PR fixes the codebase such that it works with torch 2.0.0.

However IT DOES NOT change the pip dependencies for the following reason:
     torch >= 1.13 includes functorch, installing functorch pip package breaks imports (breaks in 2.0.0).
     torch < 1.13 does not include functorch and it needs to be installed.
Do we still need to support 1.12?

### CODE REVIEW

**Comment by tancik** [comment]
Do the models still perform the same?

**Comment by jkulhanek** [comment]
I haven't tested the models' performance under the new version.

File: nerfstudio/data/utils/nerfstudio_collate.py [hunk 1]
```diff
@@ -23,7 +23,12 @@
 
 import torch
 import torch.utils.data
-from torch._six import string_classes
+
+try:
+    from torch._six import string_classes
+except ImportError:
+    # _six deprecated in torch v2
+    string_classes = (str, bytes)
```
Comment (tancik) [inline nerfstudio/data/utils/nerfstudio_collate.py:31]: Can we just edit ln 127 to `elif isinstance(elem, (str, bytes)):`?

Comment (jkulhanek) [reply nerfstudio/data/utils/nerfstudio_collate.py:31]: Sure

**Comment by adithya-Avataar** [comment]
hi @jkulhanek I was wondering if you have tried using the torch.compile feature of 2.0 with nerfstudio. I had tried using the torch.compile feature, but it was unable to build the graph. Just curious to know if you have some info on this.

**Comment by jkulhanek** [comment]
Hi, I don’t think it is possible to build the graph with tcnn and nerfacc custom cuda ops.

**Comment by jkulhanek** [comment]
On another note, nerfacto seems not to be working with torch 2.0. I get the following error: https://github.com/Lightning-AI/lightning/issues/17245

**Comment by adithya-Avataar** [comment]
Got it. Thank you

**Comment by KamiCalcium** [comment]
> 

Hi,

Did you try the workaround mentioned in that thread? I'm having the same problem for torch 2.0.

**Comment by jkulhanek** [comment]
No, lightning is not used, I was just pointing at other instances of the same error. There must be something in AMP optimisation which causes it to fail. Could the problem be caused by some gradients being computed inside of the custom tcnn kernels?

**Comment by KamiCalcium** [comment]
> No, lightning is not used, I was just pointing at other instances of the same error. There must be something in AMP optimisation which causes it to fail. Could the problem be caused by some gradients being computed inside of the custom tcnn kernels?

Could be yeah. Didn't dig too much but the easiest workaround I found is to comment out that assertion and it worked..

**Comment by gtangg12** [comment]
Few other additional things:

1) In nerfstudio/utils/profiler.py
`from torch.autograd.profiler import ContextDecorator `
ContextDecorator doesn't exist anymore. I think it may be _ContextDecorator.

2) One cannot index a cpu tensor with a cuda tensor anymore. In dataloaders, this is done (image is on cpu while image_index and its derivatives are on cuda).

**Comment by jkulhanek** [comment]
1. Sorry, I have introduced this one.
2. Wasn’t this already the case in 1.13?

**Comment by gtangg12** [comment]
Yea, I was on 1.12.1. Just a bit confused since image idx is on cuda but image cpu. Should I convert them all to cuda in the dataset?

**Comment by jkulhanek** [comment]
I think in previous torch versions if the sampled tensor was on cpu the indices were moved to cpu as well. I would think that this would be more performant then moving both to cuda as you would have to move all cached images to cuda… could be quite expensive for large datasets. I guess this very much depends on the data, but sampling on cpu should be safer.

**Comment by gtangg12** [comment]
Thanks, makes sense, thought this brings up another issue: in the dataloaders only image is excluded but mask, depth, and such are still being moved to cuda for the batch, which defeats the purpose of this ig.

**Comment by jkulhanek** [comment]
True, in another PR I wanted to address this inconsistency. I didn’t have the time to test it though.

**Comment by machenmusik** [comment]
Re: masks - see #1741?

**Comment by jkulhanek** [comment]
Depth: #1730
